pkg/nats/stream/producer_client.go

```diff
@@ -130,50 +133,30 @@ func (pc *ProducerClient) heartBeat(ctx context.Context) {
 				msg, err := pc.Conn.Request(v, data, pc.config.HeartBeatRequestTimeout)
 				if err != nil {
 					log.Ctx(ctx).Err(err).Msg("error while sending heart beat request from NATS streaming producer client")
-					nonActiveStreamIds[c] = []string{}
 					continue
 				}
 
 				var heartBeatResponse ConsumerHeartBeatResponse
 				err = json.Unmarshal(msg.Data, &heartBeatResponse)
 				if err != nil {
-					log.Ctx(ctx).Err(err).Msg("error while  parsing heart beat response from NATS streaming consumer client")
+					log.Ctx(ctx).Err(err).Msg("error while parsing heart beat response from NATS streaming consumer client")
 					continue
 				}
 
 				nonActiveStreamIds[c] = getStringList(heartBeatResponse.NonActiveStreamIds)
 			}
 
-			pc.updateActiveStreamInfo(nonActiveStreamIds)
+			pc.mu.RUnlock()
+			if len(nonActiveStreamIds) != 0 {
+				pc.updateActiveStreamInfo(nonActiveStreamIds)
+			}
```

Comment:
how about adding all this in a `defer` statement right after acquiring the lock?

Reply:
This unlock is happening in an infinite for loop. I don't think using `defer` here would be a good idea ! But let me know if you think otherwise.

Reply:
You are right. I missed the loop